src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/simple/DataStructureForStoringPlansOfSubsets.java

```diff
@@ -0,0 +1,22 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.simple;
+
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DataStructureForStoringPlansOfSubsets {
+    protected Map< Integer, Map<List<PhysicalPlan>, PhysicalPlan> > map=new HashMap<>();
+
+    public void add( final List<PhysicalPlan> subsets, final PhysicalPlan plan ) {
+        final int size=subsets.size();
+        map.putIfAbsent( size, new HashMap<>() );
+        map.get(size).put(subsets, plan);
```

Comment:
It is probably more efficient to do the `map.get(size)` right away and if the result is `null` you create the new hash map and put it into the main map.